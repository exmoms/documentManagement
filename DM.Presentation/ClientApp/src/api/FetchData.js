/**
 * @file this file includes functions that fetch data from API
 */

// Generic Delete Method
export async function deleteFromApi(url) {
  let res = await fetch(url, {
    method: "DELETE",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  });
  return res;
}

// Get MetaData Models
export async function fetchData(url) {
  let object = {};
  /*
   * send the request through a CORS proxy
   * https://stackoverflow.com/questions/43262121/trying-to-use-fetch-and-pass-in-mode-no-cors/43268098#43268098
   */
  await fetch(url, {
    method: "GET",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
  })
    .then((response) => response.json())
    .then(
      (data) => (object = data),
      (error) => console.log(error)
    );
  return object;
}
