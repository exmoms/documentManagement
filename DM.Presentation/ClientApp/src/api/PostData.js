/**
 * @file This file includes functions that help to post data to a specific API
 * @version 1.0
 */

/**
 * Post new document data.
 * Upload form-data, scanned copy of doc and attachments.
 * @author Youssef Shaaban <y.shaaban@lit-co.net>
 * @param {*} documentData
 * @param {*} images
 * @param {*} attachments
 * @param {*} url
 * @param {*} option its values either "Add" for add new document version or "Update" to update exsiting document Version
 */
export async function postDocumentData(
  documentData,
  scannedPages,
  attachments,
  url = "",
  option
) {
  var formData = new FormData();

  if (process.env.NODE_ENV === "development") {
    console.log(documentData);
    console.log(attachments);
    console.log(scannedPages);
  }

  switch (option) {
    case "Add":
      formData.append("Document", JSON.stringify(documentData));
      break;
    case "Update":
      formData.append("Version", JSON.stringify(documentData));
      break;
    default:
  }

  attachments.forEach((attachment) => {
    formData.append("Attachments", attachment);
  });
  scannedPages.forEach((scannedPage) => {
    formData.append("Scans", scannedPage);
  });

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    headers: {
      Accept: "application/json",
      // "Content-Type": "application/json",
    },
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Attachment Data.
 * Upload attachment json data + attachment file  .
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} payload
 * @param {*} attachmentFile
 * @param {*} url
 */
export async function postAttachmentData(payload, attachmentFile, url = "") {
  var formData = new FormData();

  formData.append("Attachment", JSON.stringify(payload));
  formData.append("FileAttachment", attachmentFile);

  let response = {};
  await fetch(url, {
    // content-type header should not be specified!
    method: "POST",
    body: formData,
  }).then(
    (res) => {
      response = res;
      // response.json();
    },
    (error) => console.log("[ERROR]:", error)
  );
  return response;
}

/**
 * Post Request To API
 * Send post request to api with  payload  and return a response.
 * @author Ali Daghman <ali.daghman@lit-co.net>
 * @param {*} url The api url
 * @param {*} payload js object which will be sent as request body.
 */
export async function postDataToAPI(url, payload = null, method = "POST") {

  const response = await fetch(url, {
    method: method,
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  return response;
}
