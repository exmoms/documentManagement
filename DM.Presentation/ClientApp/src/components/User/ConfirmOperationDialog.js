import React, { Component } from "react";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogContentText from "@material-ui/core/DialogContentText";
import DialogTitle from "@material-ui/core/DialogTitle";
import { withTranslation } from "react-i18next";

class ConfirmOperationDialog extends Component {
  constructor(props) {
    super(props);
    this.handleClose = this.props.handleClose;
  }
  t = this.props["t"];
  render() {
    let { type, title, contentText, buttonText } = this.props.content;
    let lang = localStorage.getItem("i18nextLng");
    return (
      <div>
        <Dialog
          open={this.props.open}
          aria-labelledby="alert-dialog-title"
          aria-describedby="alert-dialog-description"
        >
          <div
            style={
              lang === "ar" ? { textAlign: "right" } : { textAlign: "left" }
            }
          >
            <DialogTitle id="alert-dialog-title">{title}</DialogTitle>
          </div>
          <DialogContent>
            <div
              style={
                lang === "ar" ? { textAlign: "right" } : { textAlign: "left" }
              }
            >
              <DialogContentText id="alert-dialog-description">
                {contentText}
              </DialogContentText>
            </div>
          </DialogContent>
          <DialogActions>
            <Button
              id="Back"
              onClick={() => this.handleClose(false, type)}
              color="primary"
              autoFocus
            >
              {this.t("cancel")}
            </Button>
            <Button
              id="Confirm"
              type="submit"
              onClick={() => this.handleClose(true, type)}
              form="update-form"
              color="primary"
              autoFocus
            >
              {buttonText}
            </Button>
          </DialogActions>
        </Dialog>
      </div>
    );
  }
}

export default withTranslation()(ConfirmOperationDialog);
