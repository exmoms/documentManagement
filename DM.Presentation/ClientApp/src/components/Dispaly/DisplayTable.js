import LinearProgress from "@material-ui/core/LinearProgress";
import AddBox from "@material-ui/icons/AddBox";
import ArrowDownward from "@material-ui/icons/ArrowDownward";
import Check from "@material-ui/icons/Check";
import ChevronLeft from "@material-ui/icons/ChevronLeft";
import ChevronRight from "@material-ui/icons/ChevronRight";
import Clear from "@material-ui/icons/Clear";
import DeleteIcon from "@material-ui/icons/Delete";
import DeleteOutline from "@material-ui/icons/DeleteOutline";
import Edit from "@material-ui/icons/Edit";
import FilterList from "@material-ui/icons/FilterList";
import FirstPage from "@material-ui/icons/FirstPage";
import LastPage from "@material-ui/icons/LastPage";
import Remove from "@material-ui/icons/Remove";
import SaveAlt from "@material-ui/icons/SaveAlt";
import Search from "@material-ui/icons/Search";
import ViewColumn from "@material-ui/icons/ViewColumn";
import { withStyles } from "@material-ui/styles";
//import VisibilityIcon from "@material-ui/icons/Visibility";
import MaterialTable from "material-table";
import React, { forwardRef } from "react";
import { withTranslation } from "react-i18next";
//import Autocomplete from "@material-ui/lab/Autocomplete";
//import TextField from "@material-ui/core/TextField";
import { deleteFromApi } from "../../api/FetchData";
import ConfirmOperationDialog from "../User/ConfirmOperationDialog";
import { Notification } from "../Admin/Notifications";

const tableIcons = {
  Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
  Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
  Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} />),
  DetailPanel: forwardRef((props, ref) => (
    <ChevronRight {...props} ref={ref} />
  )),
  Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
  Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
  Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
  FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
  LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
  NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
  PreviousPage: forwardRef((props, ref) => (
    <ChevronLeft {...props} ref={ref} />
  )),
  ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
  Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
  SortArrow: forwardRef((props, ref) => <ArrowDownward {...props} ref={ref} />),
  ThirdStateCheck: forwardRef((props, ref) => <Remove {...props} ref={ref} />),
  ViewColumn: forwardRef((props, ref) => <ViewColumn {...props} ref={ref} />),
};

const style = (theme) => ({
  select: {
    margin: "26px",
    marginLeft: "0px",
  },
});

class DisplayTable extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false,
      id: 1,
      showConfirmOperationDialog: false,
      confirmDialogContent: {},
      classIdToDelete: -1,
      openNotifiction: false,
      error: false,
      errorMessage: [],
    };
  }

  t = this.props["t"];
  handleClose = () => {
    this.setState({ open: false });
  };

  showConfirmOperationDialog = (rowData) => {
    this.setState({
      showConfirmOperationDialog: true,
      classIdToDelete: rowData.id,
      confirmDialogContent: {
        type: "delete",
        title: this.t("dialog_deleteConfirmation"),
        contentText: this.t("dialog_documentClass_confirmDeleteDocumentClass"),
        buttonText: this.t("delete"),
      },
    });
  };

  handleCloseConfirmOperationDialog = async (ok, option) => {
    if (ok) {
      await this.deleteDocumentClass(this.state.classIdToDelete);
    } else {
      this.setState({
        showConfirmOperationDialog: false,
        entryIdToDelete: -1,
        confirmDialogContent: {},
      });
    }
  };

  deleteDocumentClass = async (id) => {
    let response = await deleteFromApi(this.props.DeleteUrl + id);
    if (response.ok) {
      this.setState(
        {
          showConfirmOperationDialog: false,
          entryIdToDelete: -1,
          confirmDialogContent: {},
          openNotifiction: true,
        },
        () => {
          this.props.onSubmit("Deleted successfully");
        }
      );
    } else {
      let json = await response.json();
      this.setState({
        showConfirmOperationDialog: false,
        entryIdToDelete: -1,
        confirmDialogContent: {},
        openNotifiction: true,
        error: true,
        errorMessage: json.error,
      });
    }
  };

  setOpenNotification = (value) => {
    this.setState({ openNotifiction: value });
  };

  render() {
    // const { classes } = this.props;
    return (
      <div style={{ maxWidth: "100%" }}>
        {this.props.loading && <LinearProgress color="secondary" />}
        <Notification
          open={this.state.openNotifiction} // must be a "state" to show notification message [true or false]
          setOpen={this.setOpenNotification} // this function changes the state "open" value.
          error={this.state.error} // if error message put [true] else for success put [false]
          errorMessage={this.state.errorMessage} // the recived error message (array of string) from server.
        />
        <MaterialTable
          title={this.props.TableTitle}
          icons={tableIcons}
          columns={this.props.Columns}
          data={this.props.Data}
          actions={[
            {
              icon: DeleteIcon,
              tooltip: "Delete",
              onClick: (event, rowData) => {
                console.log("action");
                this.showConfirmOperationDialog(rowData);
              },
            },
          ]}
          options={{
            actionsColumnIndex: -1,
            exportButton: true,
            exportAllData: true,
            exportFileName: this.props.TableTitle,
          }}
          localization={{
            pagination: {
              labelRowsSelect: this.t("rows"),
              firstAriaLabel: this.t("first_page"),
              firstTooltip: this.t("first_page"),
              previousAriaLabel: this.t("previous_page"),
              previousTooltip: this.t("previous_page"),
              nextAriaLabel: this.t("next_page"),
              nextTooltip: this.t("next_page"),
              lastAriaLabel: this.t("last_page"),
              lastTooltip: this.t("last_page"),
            },
            toolbar: {
              nRowsSelected: "{0}" + this.t("selected_row"),
              searchTooltip: this.t("search"),
              searchPlaceholder: this.t("search"),
              exportTitle: this.t("export_as_csv"),
            },
            header: {
              actions: this.t("actions"),
            },
            body: {
              emptyDataSourceMessage: this.t("no_records"),
            },
          }}
        />
        {this.state.showConfirmOperationDialog && (
          <ConfirmOperationDialog
            open={this.state.showConfirmOperationDialog}
            content={this.state.confirmDialogContent}
            handleClose={this.handleCloseConfirmOperationDialog}
          />
        )}
      </div>
    );
  }
}

export default withStyles(style)(withTranslation()(DisplayTable));
