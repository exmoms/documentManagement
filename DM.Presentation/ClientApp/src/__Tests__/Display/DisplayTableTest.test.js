import React from "react";
import ReactDom from "react-dom";
import DisplayTable from "../../components/Dispaly/DisplayTable";
import ConfirmOperationDialog from "../../components/User/ConfirmOperationDialog";
import { cleanup } from "@testing-library/react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import TextField from "@material-ui/core/TextField";
import { act } from "react-dom/test-utils";
import Button from "@material-ui/core/Button";
import * as FetchMock from "../../api/FetchData";
import { deleteFromApi } from "../../api/FetchData";
import { BrowserRouter as Router } from "react-router-dom";
import { Route, Switch } from "react-router-dom";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import { render, fireEvent } from "@testing-library/react";
import MaterialTable from "material-table";

//imports for tableIcons
import { forwardRef } from "react";
import AddBox from "@material-ui/icons/AddBox";
import ArrowDownward from "@material-ui/icons/ArrowDownward";
import Check from "@material-ui/icons/Check";
import ChevronLeft from "@material-ui/icons/ChevronLeft";
import ChevronRight from "@material-ui/icons/ChevronRight";
import Clear from "@material-ui/icons/Clear";
import DeleteOutline from "@material-ui/icons/DeleteOutline";
import Edit from "@material-ui/icons/Edit";
import FilterList from "@material-ui/icons/FilterList";
import FirstPage from "@material-ui/icons/FirstPage";
import LastPage from "@material-ui/icons/LastPage";
import Remove from "@material-ui/icons/Remove";
import SaveAlt from "@material-ui/icons/SaveAlt";
import Search from "@material-ui/icons/Search";
import DeleteIcon from "@material-ui/icons/Delete";
import ViewColumn from "@material-ui/icons/ViewColumn";
afterEach(cleanup);
describe("DisplayTable", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  test("renders without crashing", () => {
    const div = document.createElement("div");
    let columnsArray = [
      { title: "Document Class Id", field: "id" },
      { title: "Document Class Name", field: "documentClassName" },
      { title: "User Id", field: "userID" },
      { title: "Added Date", field: "addedDate" },
    ];
    let data = [
      { documentClass: "document 1" },
      { documentClass: "document 2" },
    ];
    let tableTitle = "Brows Document Classes";
    let deletedUrl = "/api/DocumentClass/DeleteDocumentClass?classId=";
    let loading = true;
    ReactDom.render(
      <DisplayTable
        Data={data}
        Columns={columnsArray}
        TableTitle={tableTitle}
        DeleteUrl={deletedUrl}
        loading={loading}
      />,
      div
    );
  });

  test("matches snapshot", () => {
    let columnsArray = [
      { title: "Document Class Id", field: "id" },
      { title: "Document Class Name", field: "documentClassName" },
      { title: "User Id", field: "userID" },
      { title: "Added Date", field: "addedDate" },
    ];
    let data = [
      { documentClass: "document 1" },
      { documentClass: "document 2" },
    ];
    let tableTitle = "Brows Document Classes";
    let deletedUrl = "/api/DocumentClass/DeleteDocumentClass?classId=";
    let loading = true;
    const wrapper = shallow(
      <DisplayTable
        Data={data}
        Columns={columnsArray}
        TableTitle={tableTitle}
        DeleteUrl={deletedUrl}
        loading={loading}
      />
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test("DisplayTable,test intilization", () => {
    let columnsArray = [
      { title: "Document Class Id", field: "id" },
      { title: "Document Class Name", field: "documentClassName" },
      { title: "User Id", field: "userID" },
      { title: "Added Date", field: "addedDate" },
    ];
    let data = [
      { documentClass: "document 1" },
      { documentClass: "document 2" },
    ];
    let tableTitle = "Brows Document Classes";
    let deletedUrl = "/api/DocumentClass/DeleteDocumentClass?classId=";
    let loading = true;
    let props = {
      Data: data,
      loading: loading,
      TableTitle: tableTitle,
      Columns: columnsArray,
      DeleteUrl: deletedUrl,
    };
    const tableIcons = {
      Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
      Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
      Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
      Delete: forwardRef((props, ref) => (
        <DeleteOutline {...props} ref={ref} />
      )),
      DetailPanel: forwardRef((props, ref) => (
        <ChevronRight {...props} ref={ref} />
      )),
      Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
      Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
      Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
      FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
      LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
      NextPage: forwardRef((props, ref) => (
        <ChevronRight {...props} ref={ref} />
      )),
      PreviousPage: forwardRef((props, ref) => (
        <ChevronLeft {...props} ref={ref} />
      )),
      ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
      Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
      SortArrow: forwardRef((props, ref) => (
        <ArrowDownward {...props} ref={ref} />
      )),
      ThirdStateCheck: forwardRef((props, ref) => (
        <Remove {...props} ref={ref} />
      )),
      ViewColumn: forwardRef((props, ref) => (
        <ViewColumn {...props} ref={ref} />
      )),
    };
    let wrapper = mount(
      <DisplayTable
        Data={data}
        Columns={columnsArray}
        TableTitle={tableTitle}
        DeleteUrl={deletedUrl}
        loading={loading}
      />
    );
    //expectations
    expect(wrapper.find(MaterialTable).props().title).toBe(tableTitle);
    expect(JSON.stringify(wrapper.find(MaterialTable).props().icons)).toBe(
      JSON.stringify(tableIcons)
    );
    expect(wrapper.find(MaterialTable).props().columns).toBe(columnsArray);
    expect(wrapper.find(MaterialTable).props().data).toBe(data);
    expect(wrapper.find(MaterialTable).props().options).toStrictEqual({
      actionsColumnIndex: -1,
      exportButton: true,
      exportAllData: true,
      exportFileName: tableTitle,
    });
  });

  test("DisplayTable, delete event", () => {
    let columnsArray = [
      { title: "Document Class Id", field: "id" },
      { title: "Document Class Name", field: "documentClassName" },
      { title: "User Id", field: "userID" },
      { title: "Added Date", field: "addedDate" },
    ];
    let data = [
      { id: 1, documentClass: "document 1" },
      { id: 2, documentClass: "document 2" },
    ];
    let tableTitle = "Brows Document Classes";
    let deletedUrl = "/api/DocumentClass/DeleteDocumentClass?classId=";
    let loading = true;
    const mockSuccessResponse = { status: 200 };
    const mockJsonPromise = Promise.resolve(mockSuccessResponse); // 2
    const fetchingMock = jest
      .spyOn(FetchMock, "deleteFromApi")
      .mockImplementation(() => mockJsonPromise);
    let wrapper = mount(
      <DisplayTable
        Data={data}
        Columns={columnsArray}
        TableTitle={tableTitle}
        DeleteUrl={deletedUrl}
        loading={loading}
      />
    );
    act(() => {
      wrapper.find(MaterialTable).props().actions[0].onClick({ target: { value: ''} } , { id: 1, documentClass :"document 1"})
    });
    wrapper.update();
    wrapper.render();
    expect(wrapper.find(ConfirmOperationDialog).length).toBe(1);
    expect(wrapper.find(ConfirmOperationDialog).at(0).props().open).toBeTruthy();
    expect(wrapper.find(ConfirmOperationDialog).at(0).props().content).toStrictEqual({
      type: "delete",
      title: "dialog_deleteConfirmation",
      contentText: "dialog_documentClass_confirmDeleteDocumentClass",
      buttonText: "delete",
    });
  });
});
