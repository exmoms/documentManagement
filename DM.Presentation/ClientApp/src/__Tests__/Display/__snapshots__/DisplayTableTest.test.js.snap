// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayTable matches snapshot 1`] = `
<withI18nextTranslation(DisplayTable)
  Columns={
    Array [
      Object {
        "field": "id",
        "title": "Document Class Id",
      },
      Object {
        "field": "documentClassName",
        "title": "Document Class Name",
      },
      Object {
        "field": "userID",
        "title": "User Id",
      },
      Object {
        "field": "addedDate",
        "title": "Added Date",
      },
    ]
  }
  Data={
    Array [
      Object {
        "documentClass": "document 1",
      },
      Object {
        "documentClass": "document 2",
      },
    ]
  }
  DeleteUrl="/api/DocumentClass/DeleteDocumentClass?classId="
  TableTitle="Brows Document Classes"
  classes={
    Object {
      "select": "withI18nextTranslation(DisplayTable)-select-1",
    }
  }
  loading={true}
/>
`;
